File: PigLatin.js
//var str = "HeLLo World! I can't wait to explore your VAST forests. The-End!";
var str = prompt("Enter a string to be translated to Pig Latin: ");

var wordArray = [];
wordArray = str.split(" ");

var output = pigLatin(wordArray);
console.log(output);

function pigLatin(wordArray) {
    var finalString = ""; //translated string
    for (var i = 0; i < wordArray.length; i++) { //for every word in sentence... translate each word
        var cStack = []; //stores indices of capitalized letters
        var exclamationStack = []; //stores indices of exclamation points
        var questionMarkStack = []; //stores indices of question marks
        var periodStack = []; //stores indices of periods
        var apostStack = []; //stores indices of apostrophes
        var word = []; //the single word being analyzed

        word = wordArray[i];

        //format check
        if (word.length == 1) { //checks if it's a single letter
            finalString += word + " ";
            continue;
        }
        if (word.toLowerCase() == "way") { //checks if it is just "way"
            finalString += word + " ";
            continue;
        }

        if (word.search("-") != -1) { //checks if it has a hyphen
            word = hyphen(word); //calls hyphen function which will call pigLatin function
            finalString += word + " ";
            continue;
        }

        for (var j = 0; j < word.length; j++) { //Scan through word for punctuation and capitalization
            //Store indices in stacks and take out punct & caps
            if (word[j] == "!") {
                exclamationStack.push(j);
                word = word.replace("!", "");
                var origWord = word;
            }
            if (word[j] == "?") {
                questionMarkStack.push(j);
                word = word.replace("?", "");
                var origWord = word;
            }
            if (word[j] == ".") {
                periodStack.push(j);
                word = word.replace(".", "");
                var origWord = word;
            }
            if (word[j] == "'") {
                apostStack.push(j);
                word = word.replace("'", "");
                var origWord = word;
            }
            if (word[j] == null) {
                continue;
            }
            if (word[j] == word[j].toUpperCase()) //if capitalized, store index into stack
                cStack.push(j);
        }

        //turn everything into regular text without caps in order to add proper capitalization later
        word = word.toLowerCase();

        //Two ways to translate word into pig latin... depending if it's a vowel or consonant

        //if word starts with a vowel...
        if (word[0] == 'a' || word[0] == 'e' || word[0] == 'i' || word[0] == 'o' || word[0] == 'u') {
            var ending = "way";
            word += ending;
        } else { //if word starts with a consonant
            var ending = word.substr(0, 1) + "ay";
            word = word.substr(1, word.length);
            word += ending;
        }

        var clength = cStack.length; //calls capitalization function to keep proper capitalization
        word = capitalize(word, cStack, clength);

        //adding punctuation back to the word
        for (var y = 0, length = periodStack.length; y < length; y++) {
            var pos = periodStack.pop(); //using stacks to add punctuation back
            var difference = origWord.length - pos; //comparing where the punctuation is relative to end of word
            pos = word.length - difference; //to calculate where it should be in translated word
            word = word.substr(0, pos) + "." + word.substr(pos);
        }
        for (var y = 0, length = exclamationStack.length; y < length; y++) {
            var pos = exclamationStack.pop();
            var difference = origWord.length - pos;
            pos = word.length - difference;
            word = word.substr(0, pos) + "!" + word.substr(pos);
        }
        for (var y = 0, length = questionMarkStack.length; y < length; y++) {
            var pos = questionMarkStack.pop();
            var difference = origWord.length - pos;
            pos = word.length - difference;
            word = word.substr(0, pos) + "?" + word.substr(pos);
        }
        for (var y = 0, length = apostStack.length; y < length; y++) {
            var pos = apostStack.pop();
            var difference = origWord.length - pos;
            pos = word.length - difference;
            word = word.substr(0, pos) + "'" + word.substr(pos);
        }

        finalString += word + " ";
    }
    return finalString;
}


function capitalize(word, stack, length) { //capitalizes the word and returns
    for (var i = 0; i < length; i++) {
        var pos = stack.pop();
        if (pos == 0) { //if first letter is uppercase
            word = word.substr(0, 1).toUpperCase() + word.substr(1);
        }
        if (pos == word.length - 1) { //if last letter is uppercase
            word = word.substr(0, word.length - 1) + word.substr(word.length - 1).toUpperCase();
        }
        if (pos > 0 && pos < word.length - 1) { //if any letters in the middle are uppercased
            word = word.substr(0, pos) + word[pos].toUpperCase() + word.substr(pos + 1);
        }
    }
    return word;
}

function hyphen(word) { //deals with words that have hyphens to be dealt with as two separate words
    var pos = word.search("-");
    var firstWord = word.substr(0, pos); //separating the two words
    var secondWord = word.substr(pos + 1);
    var arr = [];
    arr.push(firstWord);
    arr.push(secondWord);
    var finalWord = pigLatin(arr); //call pigLatin function to translate
    finalWord = finalWord.replace(" ", "-"); //puts words back together
    return finalWord;

}

File: PigLatin.rb
# SENTENCE SPLITTER

  def wordify(sentence)
    return sentence_array = sentence.split(" ")
  end

# WORD DETERMINER (#Will the word be transformed? And if so, how?)

  def determine_and_translate(word)
    vowels = ['a','e', 'i', 'o', 'u', 'A', 'E', 'I', 'O', 'U']
    if word.length == 0 || word.length == 1
      return word
    elsif ends_with_way?(word)
      return word
    elsif vowels.include? word[0]
      return change_vowel_starters(word)
    else
      return change_consonant_starters(word)
    end
    return word
  end

# ACTIONS TO FOLLOW DEPENDING ON TYPE OF WORD

  def change_consonant_starters(word)
    word_array = word.downcase.split("")
    first_letter = word_array.shift
    if /\W/.match(word_array[-1]).class == MatchData
      final_word = word_array.insert(-2, first_letter+"ay").join
    else
      word_array << first_letter
      final_word = word_array.join + "ay"
    end
    if word[0] && word[0].capitalize === word[0]
      return final_word.capitalize!
    else
      return final_word
    end
  end

  def change_vowel_starters(word)
    word_array = word.split("")
    if /\W/.match(word_array[-1]).class == MatchData
      final_word = word_array.insert(-2, "way").join
    else
      final_word = word + "way"
    end
  end

# HELPER METHODS

  def has_punctuation?
    true if word.include(/[[:punct:]]/)
    # /\W/.match(word_array[-1]).class == MatchData
  end

  def has_apostrophe?
    true if word.include("'")
  end

  def ends_with_way?(word)
    word[-3..-1] == "way"
  end

# FINAL METHOD

def pig_latin(sentence)
  words = wordify(sentence)
  final_sentence = []
  words.each do |word|
    if word.include? '-'
      first,second=word.split("-")
      latinified = "#{determine_and_translate(first)}-#{determine_and_translate(second)}"
    else
      latinified = determine_and_translate(word)
    end
  final_sentence << latinified
  end
  return final_sentence.join(" ")
end

# USER INTERFACE

puts "Welcome to the Pig Latin Translator! Please enter a word or sentence you would like to translate. Type '1' to exit:"
input = gets.chomp
until input == "1"
    puts pig_latin(input)
    puts "Enter another word or sentence to tranlate or type '1' to exit: "
    input = gets.chomp
end




# TEST CASES BELOW


# sentence = "HeLLo World! I can't wait to explore your VAST forests. The-End!"
# puts determine_and_translate("Hello!")
# puts change_consonant_starters("Hello!")
# puts change_vowel_starters("Apple!")
# puts change_consonant_starters("Can't")
# puts pig_latin(sentence)
# puts pig_latin("Callaway")
